hbase-server/src/main/java/org/apache/hadoop/hbase/procedure2/store/region/RegionFlusherAndCompactor.java

```diff
@@ -136,9 +156,43 @@ static void setupConf(Configuration conf) {
       flushPerChanges, flushIntervalMs);
   }
 
+  private boolean isHFileDeleteable(FileStatus status) {
+    long currentTime = EnvironmentEdgeManager.currentTime();
+    long time = status.getModificationTime();
+    long life = currentTime - time;
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("HFile life: {}ms, ttl: {}ms, current: {}, from: {}", life, compactedHFileTTLMs,
+        currentTime, time);
+    }
+    if (life < 0) {
+      LOG.warn("Found a hfile ({}) newer than current time ({} < {}), probably a clock skew",
+        status.getPath(), currentTime, time);
+      return false;
+    }
+    return life > compactedHFileTTLMs;
+  }
+
+  void cleanupCompactedHFiles() throws IOException {
+    HStore store = Iterables.getOnlyElement(region.getStores());
+    // our HFiles are on the WAL file system, but the global HFile archive directory is on the
+    // root(HFile) file system, we can not move between two different file systems. So here we have
+    // to implement our own TTL cleaner.
+    Path archiveDir = HFileArchiveUtil.getStoreArchivePath(conf, region.getRegionInfo(),
+      store.getColumnFamilyDescriptor().getName());
+    FileSystem fs = archiveDir.getFileSystem(conf);
+
+    for (FileStatus status : fs.listStatus(archiveDir)) {
+      if (isHFileDeleteable(status)) {
```

Comment:
We should be asking the other cleaner delegates if it's okay to delete these files. I could see it being either the WAL cleaner delegates or the HFile Cleaner delegates, but it needs to be one of them.

Reply:
We are on a different file system comparing to other HFiles so it is not suitable to ask the HFileCleaner, and it seems strange to test a HFile with WALCleaner?

Reply:
Then we should instantiate our own hfile cleaner delegate chain. Defualt it to the TTL delegate and we can skip a custom impl here as well.

Reply:
Maybe a bit overkill? AFAIK, the only suitable HFileCleaner for this region is the TTL one, others like Links or Back Ref are not needed. Just saying. If you think this is the correct way then I could pass the ChoreService in and create a cleaner chain for cleaning the hfiles.

Reply:
No not overkill. This is the specific mechanism I mentioned needing to feel comfortable with this landing in branches where I will probably have to support deployments.

I agree that the TTL cleaner is the only one needed by default.

At some point something will go wrong where I need to avoid things getting cleaned out of archive. Being able to reuse a cleaner delegate that normally is used with HFiles will make that easier. It's exactly the kind of operational reuse that makes this implementation shift attractive.